projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad-task-definitions.php

```diff
@@ -424,34 +424,92 @@ function wpcom_launchpad_track_completed_task( $task_id, $extra_props = array()
+	}
+
+	return $mapping;
+}
 /**
  * Mark a task as complete.
  *
- * @param string $task_id The task ID.
+ * @param string|string[] $task_ids The task IDs that should be marked as complete.
  * @return bool True if the task was marked as complete, false otherwise.
  */
-function wpcom_mark_launchpad_task_complete( $task_id ) {
+function wpcom_mark_launchpad_task_complete( $task_ids ) {
+	$task_ids = is_array( $task_ids ) ? $task_ids : array( $task_ids );
+	$values   = array_fill_keys( $task_ids, true );
+
+	$result = wpcom_launchpad_update_task_status( $values );
+
+	return $result;
+}
+
+/**
+ * Mark a task as incomplete. This method is here for backwards compatibility of
+ * the Rest API that allowed marking a task as incomplete.
+ *
+ * @param string|string[] $task_ids The task IDs that should be marked as complete.
+ * @return bool True if the task was marked as complete, false otherwise.
+ */
+function wpcom_mark_launchpad_task_incomplete( $task_ids ) {
+	$task_ids = is_array( $task_ids ) ? $task_ids : array( $task_ids );
+	$values   = array_fill_keys( $task_ids, false );
+
+	return wpcom_launchpad_update_task_status( $values );
+}
+
+/**
+ * Updates task/s statuses. If a non-existent task is passed, it will be ignored.
+ * It will trigger the completion event in Tracks for the tasks that were marked as complete.
+ *
+ * @param array $new_statuses Array of mappings [ task_id: string => status: bool ].
+ * @return bool True if the option where the statuses are stored was updated, false otherwise.
+ */
+function wpcom_launchpad_update_task_status( $new_statuses ) {
 	$task_definitions = wpcom_launchpad_get_task_definitions();
+	$reverse_id_map   = wpcom_launchpad_get_reverse_id_mappings();
 
-	// If the task ID isn't defined, return false.
-	if ( ! isset( $task_definitions[ $task_id ] ) ) {
-		return false;
-	}
+	$statuses               = array();
+	$original_status_values = (array) get_option( 'launchpad_checklist_tasks_statuses', array() );
 
-	// If the task has an id_map, use that instead.
-	$key = $task_id;
-	if ( isset( $task_definitions[ $task_id ]['id_map'] ) ) {
-		$key = $task_definitions[ $task_id ]['id_map'];
-	}
+	foreach ( $new_statuses as $task_id => $value ) {
+		// Find out wether the task id sent is the actual task id or the id_map.
+		$actual_task_id = ( isset( $reverse_id_map[ $task_id ] ) ) ? $reverse_id_map[ $task_id ] : $task_id;
+		if ( ! isset( $task_definitions[ $actual_task_id ] ) ) {
+			// Bail if no task definition was found.
+			continue;
+		}
 
-	$statuses         = get_option( 'launchpad_checklist_tasks_statuses', array() );
-	$statuses[ $key ] = true;
-	$result           = update_option( 'launchpad_checklist_tasks_statuses', $statuses );
+		// If the task has an id_map, use that, otherwise use the actual task id as the key.
+		$task                        = $task_definitions[ $actual_task_id ];
+		$stored_task_id              = isset( $task['id_map'] ) ? $task['id_map'] : $actual_task_id;
```

Comment:
I am not sure why we're not just using `$actual_task_id` here. This either needs a comment to explain why the second lookup is necessary, or we need to use the already-resolved `$actual_task_id`.

Reply:
TLDR; I've made this change to prevent changing the API response when being called with an `id_map` instead of a `task_id`.

[This preexisting test](https://github.com/Automattic/jetpack/blob/trunk/projects/packages/jetpack-mu-wpcom/tests/php/features/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-launchpad-test.php#L91) implied that the API might be called with `id_maps` directly. This use case was supported by the API endpoint because it was storing directly whatever it was being sent, but now, since we are making everything go through the `wpcom_mark_launchpad_task_complete` function I had to make this changes.
The idea behind this code is that task_ids and id_maps can be used interchangeably, but the value stored in the option is always the `id_map`.
 
If we used the `actual_task_id` as key then that test will need to be changed [here](https://github.com/Automattic/jetpack/blob/trunk/projects/packages/jetpack-mu-wpcom/tests/php/features/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-launchpad-test.php#L106) to expect `domain_upsell` key instead of `domain_upsell_deferred`. While making this change is easier it would be changing the rest API response.

Reply:
Apologies for the delay in getting back to you on this point.

At this stage, I am not sure what we are looking up to find `$actual_task_id`. As far as I am concerned, there are two relevant values: the requested task id, and the task option key. In the logic we have, we would end up ignoring the requested task ID in cases where another task maps to that ID, but the request was for the underlying task. (A current example of this is for `first_post_published` and `first_post_published_newsletter`.

So I think we should either use the requested task ID or the stored task ID as the returned key for that item. I would strongly lean towards using the requested task ID, as the internal storage should be irrelevant to the client.

To lay this out in a more concrete way, how does the following look?

```suggestion
	foreach ( $new_statuses as $requested_task_id => $new_value ) {
		// If the requested task is defined, resolve the right underlying task ID.
		if ( isset( $task_definitions[ $requested_task_id ] ) ) {
			// If the task is defined, we either need to use the `id_map`, or the task itself.
			$resolved_task_id = isset( $task_definitions[ $requested_task_id ]['id_map'] ) ? $task_definitions[ $requested_task_id ]['id_map'] : $requested_task_id;
		} elseif ( isset( $reverse_id_map[ $requested_task_id ] ) ) {
			// Otherwise we need to check if the task only has an ID map.
			// NOTE: This may not be possible, or might not be something we want to allow via the API. I think we only want clients to make API calls for true external tasks.
			$resolved_task_id = $reverse_id_map[ $requested_task_id ];
		} else {
			continue;
		}

		// $resolved_task_id is what we should use for storage, and $requested_task_id is what we'd return to the client.
```

Reply:
I agree with the proposed changes, I think this sums them up:
* you can call the function with a task id (a key in the `$task_definitions` array) or with a id_map. 
* The task id will take precedence over id_map.
* `$resolved_task_id` is going to be used for storage

The only doubt I have with the suggested code is in this part
<img width="710" alt="image" src="https://github.com/Automattic/jetpack/assets/375980/2f6fabef-e7bf-4e96-9315-53a8d8b84a0a">
Suppose someone calls the function with the task `domain_upsell_deferred` (which is an id_map). `$resolved_task` is going to be defined in line `534` as `domain_upsell`. But if someone calls the function with the actual task id (`domain_upsell`) `$resolved_task_id` is going to have the `domain_upsell_deferred` value. 
In the end the function will change it's storage location for the same task based on wether a task_id or an id_map is provided. Shouldn't line 534 be: `$resolved_task_id = $requested_task_id`?

Reply:
I have also added some explicit test cases in https://github.com/Automattic/jetpack/pull/31471/commits/e12bf3c2ba3492d7dc1cf0ec451d72a2587ad0f1.